packages/snaps-controllers/src/services/offscreen/OffscreenExecutionService.ts

```diff
@@ -0,0 +1,110 @@
+import { BrowserRuntimePostMessageStream } from '@metamask/post-message-stream';
+import { nanoid } from 'nanoid';
+
+import {
+  AbstractExecutionService,
+  ExecutionServiceArgs,
+  Job,
+} from '../AbstractExecutionService';
+import { OffscreenPostMessageStream } from './OffscreenPostMessageStream';
+
+type OffscreenExecutionEnvironmentServiceArgs = {
+  documentUrl: URL;
+  frameUrl: URL;
+} & ExecutionServiceArgs;
+
+export class OffscreenExecutionService extends AbstractExecutionService<string> {
+  public readonly documentUrl: URL;
+
+  public readonly frameUrl: URL;
+
+  readonly #runtimeStream: BrowserRuntimePostMessageStream;
+
+  /**
+   * Create a new offscreen execution service.
+   *
+   * @param args - The constructor arguments.
+   * @param args.documentUrl - The URL of the offscreen document to use as the
+   * execution environment. This must be a URL relative to the location where
+   * this is called. This cannot be a public (http(s)) URL.
+   * @param args.frameUrl - The URL of the iframe to load inside the offscreen
+   * document.
+   * @param args.messenger - The messenger to use for communication with the
+   * `SnapController`.
+   * @param args.setupSnapProvider - The function to use to set up the snap
+   * provider.
+   */
+  constructor({
+    documentUrl,
+    frameUrl,
+    messenger,
+    setupSnapProvider,
+  }: OffscreenExecutionEnvironmentServiceArgs) {
+    super({
+      messenger,
+      setupSnapProvider,
+    });
+
+    this.documentUrl = documentUrl;
+    this.frameUrl = frameUrl;
+    this.#runtimeStream = new BrowserRuntimePostMessageStream({
+      name: 'parent',
+      target: 'child',
+    });
+  }
+
+  /**
+   * Send a termination command to the offscreen document.
+   *
+   * @param job - The job to terminate.
+   */
+  protected async terminateJob(job: Job<string>) {
+    // The `AbstractExecutionService` will have already closed the job stream,
+    // so we write to the runtime stream directly.
+    this.#runtimeStream.write({
```

Comment:
Could this not be handled in the `OffscreenSnapExecutor` when `terminate` is received?

Reply:
We need this for graceful termination. `terminateJob` actually removes the iframe, while the `terminate` JSON-RPC method performs termination inside the iframe.

Reply:
I know. I was wondering if we could handle the removal of the iframe in the executor when receiving the RPC method to terminate. However, I'm realizing that it wouldn't be possible as that is an async operation